File: typescript/tracing/together-ai/07_expected_error.ts
import {
  captureFeature,
  createRespan,
  createTogether,
  logExampleResult,
  runWithTogetherWorkflow,
  shutdownRespan,
} from "./_shared.js";

const workflowName = "together-ai-ts-expected-error";
const respan = createRespan("together-ai-typescript-expected-error");

try {
  const details = await runWithTogetherWorkflow(respan, workflowName, async () => {
    const client = createTogether();
    const result = await captureFeature(workflowName, async () =>
      await client.chat.completions.create({
        model: "respan/nonexistent-together-model",
        messages: [{ role: "user", content: "This request is expected to fail." }],
        max_tokens: 8,
      }),
    );
    return { expectedError: !result.ok, result };
  });

  logExampleResult(workflowName, details);
} finally {
  await shutdownRespan(respan);
}

File: python/tracing/openrouter/07_expected_error.py
"""Deterministic OpenRouter-compatible 429 failure coverage."""

from __future__ import annotations

from _shared import close_sync, make_mock_client, make_respan
from openai import RateLimitError
from respan import workflow


def main() -> None:
    respan = None
    client = None
    try:
        respan = make_respan(scenario="expected_429")
        client, model = make_mock_client()

        @workflow(name="openrouter_expected_429")
        def run(trigger_prompt: str) -> str:
            try:
                client.chat.completions.create(
                    model=model,
                    messages=[{"role": "user", "content": trigger_prompt}],
                )
            except RateLimitError as exc:
                if exc.status_code != 429:
                    raise AssertionError(
                        f"expected HTTP 429, received {exc.status_code}"
                    ) from exc
                return "Observed expected deterministic OpenRouter 429"
            raise AssertionError(
                "expected the deterministic OpenRouter request to fail"
            )

        print(run("trigger deterministic 429"))
    finally:
        close_sync(respan=respan, client=client)


if __name__ == "__main__":
    main()
